### PR TITLE
Updated decode function with advanced features and added dummy model Enhanced decode function with softmax toggle, top-k, and top-p filtering.…

### PR DESCRIPTION
**Pull request by Rahulchaube1**
Summary:
- Updated the `decode()` function to support softmax toggling, top-k sampling, and top-p (nucleus) sampling.
- Added a dummy transformer model to test the decode function and ensure it works with the new features.

Key Changes:
- `decode()` function enhancements:
  - Added softmax toggle for logit manipulation.
  - Implemented top-k and top-p filtering for better sampling.
  - Integrated a dummy model to test and validate the `decode()` function.

### CODE REVIEW

**Review by KAVYANSHTYAGI — COMMENTED** [review]
Suggestions 
Automated Assertions:
Replace print() with simple checks like:

assert logits_or_probs.shape[-1] == vocab_size  # Replace with actual size if known
assert torch.all(logits_or_probs >= 0)  # If softmax applied
This would make the test self-validating.

Add CPU/GPU Flex Toggle:
It's great that device='cpu' is used, but a dynamic device selector like:

device = 'cuda' if torch.cuda.is_available() else 'cpu'
helps avoid future hardcoded issues.

Filename Suggestion:
test_kernel.py is a bit generic. Something like test_decode_sampling.py might better reflect the intent.

Move to tests/ folder if available:
To keep inference/ clean, consider relocating this file under a tests/ or examples/ directory, if one exists.
